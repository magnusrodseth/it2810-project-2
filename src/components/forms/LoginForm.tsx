import Button from '../../components/common/Button'
import { useState } from 'react'
import { isAuthenticated } from '../../api/auth'
import useFeedback from '../../hooks/forms/useFeedback'
import { mapStatusToMessage } from '../../utils/api/handlers'
import { AxiosError } from 'axios'
import { AuthenticationResponse } from '../../types/api'
import Feedback from './Feedback'
import classNames from '../../utils/common/classNames'
import InputField from './InputField'
import { useRedirectIfAuthenticated } from '../../hooks/auth/useRedirectIfAuthenticated'
import isValidProjectUrl from '../../utils/validators/isValidProjectUrl'
import useSetCredentials from '../../hooks/forms/useSetCredentials'

const LoginForm = () => {
  const [feedback, severity, setFeedback] = useFeedback()
  const [url, setUrl] = useState('')
  const [accessToken, setAccessToken] = useState('')

  useRedirectIfAuthenticated()

  const handleLogin = async () => {
    if (!isValidProjectUrl(url)) {
      setFeedback('Invalid project URL', 'error')
      return
    }

    await isAuthenticated(url, accessToken)
      .then((response) => {
        setFeedback(response.message, 'success')
      })
      .catch((error: AxiosError<AuthenticationResponse>) => {
        const feedback =
          error.response?.data.message || mapStatusToMessage(error.response?.status || 500)

        setFeedback(feedback, 'error')
      })
  }

  const isValid = severity === 'success'
  useSetCredentials(isValid, url, accessToken)

  return (
    <div>
      {feedback && (
        <div className="my-4 text-center">
          <Feedback feedback={feedback} severity={severity} />
        </div>
      )}

      <form
        className={classNames(
          'flex flex-col justify-center',
          'rounded-lg rise-on-hover bg-base-300 shadow-md hover:shadow-lg',
          'space-y-8 p-6 md:p-10'
        )}
        onSubmit={(event) => {
          event.preventDefault()
          handleLogin()
        }}>
        <InputField
          label="GitLab Repository URL"
          id="input-url"
          placeholder="Type here"
          dataTestId="input-url"
          onChange={(event) => setUrl(event.target.value)}
          required
        />

        <InputField
          label="Personal access token"
          id="input-access-token"
          placeholder="Type here"
          dataTestId="input-access-token"
          onChange={(event) => setAccessToken(event.target.value)}
          required
        />

        <Button color="secondary" type="submit" className="mt-6">
          Sign In
        </Button>
      </form>
    </div>
  )
}

export default LoginForm
