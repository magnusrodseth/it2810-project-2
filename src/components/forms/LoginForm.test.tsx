import renderer, { act } from 'react-test-renderer'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import LoginForm from './LoginForm'

describe('LoginForm', () => {
  beforeAll(() => {
    window.location.hash = '#/login'
  })

  it('renders correctly', () => {
    const component = renderer.create(<LoginForm />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('requires input data', () => {
    render(<LoginForm />)
    const urlInput = screen.getByTestId('input-url')
    const accessTokenInput = screen.getByTestId('input-access-token')

    expect(urlInput).toBeRequired()
    expect(accessTokenInput).toBeRequired()

    // Enter invalid input and expect feedback
    act(() => {
      fireEvent.change(urlInput, { target: { value: 'invalid input' } })
      fireEvent.change(urlInput, { target: { value: 'invalid input' } })

      fireEvent.click(screen.getByRole('button'))
      waitFor(() => expect(screen.getByTestId('feedback')).toBeInTheDocument())
    })
  })
})

export { }
