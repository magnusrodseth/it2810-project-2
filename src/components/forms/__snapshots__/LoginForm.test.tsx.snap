// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders correctly 1`] = `
<div>
  <form
    className="flex flex-col justify-center rounded-lg rise-on-hover bg-base-300 shadow-md hover:shadow-lg space-y-8 p-6 md:p-10"
    onSubmit={[Function]}
  >
    <div
      className="flex flex-col justify-center items-start mx-auto space-y-4"
    >
      <label
        className="font-bold"
        htmlFor="input-url"
      >
        GitLab Repository URL
      </label>
      <input
        className="input input-secondary focus:outline-offset-0 input-bordered"
        data-testid="input-url"
        id="input-url"
        onChange={[Function]}
        placeholder="Type here"
        required={true}
        type="text"
      />
    </div>
    <div
      className="flex flex-col justify-center items-start mx-auto space-y-4"
    >
      <label
        className="font-bold"
        htmlFor="input-access-token"
      >
        Personal access token
      </label>
      <input
        className="input input-secondary focus:outline-offset-0 input-bordered"
        data-testid="input-access-token"
        id="input-access-token"
        onChange={[Function]}
        placeholder="Type here"
        required={true}
        type="text"
      />
    </div>
    <button
      className="btn py-2 px-4 mt-6 btn-secondary"
    >
      Sign In
    </button>
  </form>
</div>
`;
